elastic-agent-client.proto

```diff
@@ -112,15 +114,301 @@ message ActionResponse {
+// as `STARTING` and send a new `UnitExpected` for the missing unit.
+message CheckinObserved {
+    // Token that is used to uniquely identify the connection to the Elastic Agent.
+    string token = 1;
+    // Units observed state.
+    repeated UnitObserved units = 2;
+}
+
 service ElasticAgent {
     // Called by the client to provide the Elastic Agent the state of the application.
     //
     // A `StateObserved` must be streamed at least every 30 seconds or it will result in the
-    // application be automatically marked as FAILED, and after 60 seconds it will be force killed and
-    // restarted.
+    // application be automatically marked as FAILED, and after 60 seconds it will force kill the entire
+    // process and restart it.
     rpc Checkin(stream StateObserved) returns (stream StateExpected);
 
+    // Called by the client to provide the Elastic Agent the state of the application over the V2 protocol.
+    //
+    // A `CheckinObserved` must be streamed at least every 30 seconds or it will result in the
+    // set of units automatically marked as FAILED, and after 60 seconds it will force kill the entire
+    // process and restart it.
+    rpc CheckinV2(stream CheckinObserved) returns (stream CheckinExpected);
+
     // Called by the client on connection to the GRPC allowing the Elastic Agent to stream action
     // requests to the application and the application stream back responses to those requests.
+    //
+    // Request and response is swapped here because the Elastic Agent sends the requests in a stream
+    // to the connected process. The order of response from the process does not matter, it is acceptable
+    // for the response order to be different then the request order.
     rpc Actions(stream ActionResponse) returns (stream ActionRequest);
 }
+
+// Type of transaction to start.
+enum StoreTxnType {
+    READ_ONLY = 0;
+    READ_WRITE = 1;
+}
+
+// Begins a new transaction.
+//
+// A started transaction must either have commit or discard called.
+message StoreBeginTxnRequest {
+    // Token that is used to uniquely identify the connection to the Elastic Agent.
+    string token = 1;
+    // ID of the unit.
+    string unit_id = 2;
+    // Type of transaction to start.
+    StoreTxnType type = 3;
+}
+
+// Response for a started transaction.
+message StoreBeginTxnResponse {
+    // Transaction ID.
+    string id = 1;
+}
+
+// Gets a key from the store.
+message StoreGetKeyRequest {
+    // Token that is used to uniquely identify the connection to the Elastic Agent.
+    string token = 1;
+    // Transaction ID.
+    string txn_id = 2;
+    // Name of the key.
+    string name = 3;
+}
+
+// Response of the retrieved key.
+message StoreGetKeyResponse {
+    // Status result of the get.
+    enum Status {
+        // Action was successful.
+        FOUND = 0;
+        // Action has failed.
+        NOT_FOUND = 1;
+    }
+    Status status = 1;
```

Comment:
Could you just communicate this information using the error returned by the RPC call?

There are built in status codes you could use for this. `OK` means found, `NOT_FOUND` means not found: https://grpc.github.io/grpc/core/md_doc_statuscodes.html

Reply:
I have mixed feelings on this, I believe we could use the predefined code. For a locally K/V store that would make sense. Would these return code be OK or prefixed enough if we move to an external K/V Storage like Elasticsearch?

Reply:
The errors from RPC calls in Go will have both a code and a string error message, see https://jbrandhorst.com/post/grpc-errors/ for examples.

The point is to allow easily handling the standardized codes as standard errors in clients (like HTTP status codes), but providing an error message to add context.

I think the codes are doing the same job as your enum here. They are widely used in the gRPC ecosystem.

Reply:
You convince me, here I don't have an argument agains't it.